src/sentry/tasks/integrations.py

```diff
@@ -51,3 +50,41 @@ def sync_assignee_outbound(external_issue_id, user_id, assign, **kwargs):
     else:
         user = User.objects.get(id=user_id)
     integration.get_installation().sync_assignee_outbound(external_issue, user, assign=assign)
+
+
+@instrumented_task(
+    name='sentry.tasks.integrations.sync_status_outbound',
+    queue='integrations',
+    default_retry_delay=60 * 5,
+    max_retries=5
+)
+@retry(exclude=(ExternalIssue.DoesNotExist, Integration.DoesNotExist))
+def sync_status_outbound(external_issue_id, is_resolved, **kwargs):
+    external_issue = ExternalIssue.objects.get(id=external_issue_id)
+    integration = Integration.objects.get(id=external_issue.integration_id)
+    integration.get_installation().sync_status_outbound(external_issue, is_resolved)
+
+
+@instrumented_task(
+    name='sentry.tasks.integrations.kick_off_status_syncs',
+    queue='integrations',
+    default_retry_delay=60 * 5,
+    max_retries=5
+)
+@retry()
+def kick_off_status_syncs(project_id, group_id, is_resolved, **kwargs):
```

Comment:
So the only concern I have here is some potential order of operations issues and the `is_resolved` status being out of sync.

There are multiple places where this task gets triggered, and this task happens asynchronously. So it's possible that say, on a worst case, this gets run hours later. (Unlikely, but exaggerating to make it more obvious). What happens if the status has changed between the time the task was triggered and the time the task runs?

What happens if two tasks are triggered around the same time, one for is_resolved = True and the other for is_resolved = False, but they get executed in the wrong order?

There are no guarantees of lineralizability here with how we run tasks, so it's possible that two could be executed in the wrong order.

I'd suggest, if we care about this, querying back the status within the task itself. Potentially inside of the `sync_status_outbound` task as well, and query the group's status as close to setting it as possible. That way we can be as close to accurate as possible.

One other thing is you might want to introduce a lock around updating an issue for a specific group to avoid a race condition with two tasks trying to mutate the same one with potentially conflicting values. That's likely less important if you check the state when syncing upstream.

Reply:
good call. 

ok so maybe the best thing to do is add something to both tasks like:
```python
try:
    group_status = Group.objects.filter(id=group_id).values_list('status')[0]
except IndexError:
    return
correct_status = GroupStatus.RESOLVED if is_resolved else GroupStatus.UNRESOLVED
if group_status != correct_status:
    return
```

Reply:
I mean, why not just say `group_status = Group.objects.filter(id=group_id).values_list('status', flat=True)[0]` ?

Why do you need to send it along at all to compare it?

Reply:
so i still think that could be valuable in determining if we even need to sync? like if we're triggering the task because someone marked as resolved, but then immediately undoes that, we could avoid actually trying to sync anything. does that make sense?

Reply:
I don't think that'd be as easy to do as it sounds since it'd require double checking with the backend. I think in this rare case, it'd be fine to blast two of the same thing to be safe.

Reply:
by the backend do you mean the external integration? I was basically thinking this:

Say a user sets to resolved and then immediately sets to unresolved:
1. If we compare `is_resolved` and `Group.status` two tasks will be fired, but only the one where is_resolved=False will result in an external api call
2. if we don't compare and just use the current group status, we'll just get two tasks and both will result in external api calls (with the correct statuses)

I'll change to not pass it, but I'm interested in trying to understand why you think that's better. Is it just because it's not worth the complication of passing it and comparing? Or is it that you think it's possible option will break down and still result in incorrect statuses?

Reply:
Yup, your assessment is right. I think that's a fine optimization. My statement was just that I don't think it's big enough deal to care about since that case would pretty rare I'd imagine. But I don't think it hurts to do the comparison.

Ideally, you'd compare it against what the external integration state is since that's the canonical truth, but can't do that efficiently.

Reply:
ok cool. updated to just check group status in the task that actually does the syncing. thx for your help!